convokit/storage/storageManager.py

```diff
@@ -0,0 +1,214 @@
+from typing import List, Collection, Callable, Set, Generator, Tuple, Optional, ValuesView, Union, MutableMapping
+from pymongo import MongoClient
+from pymongo.database import Database
+from random import randrange
+import os
+import uuid
+from yaml import load, dump, Loader, Dumper
+
+from .dbMappings import DBCollectionMapping, DBDocumentMapping
+from .memMappings import MemCollectionMapping, MemDocumentMapping
+from convokit.util import warn
+from .convoKitIndex import ConvoKitIndex
+
+TEXT = ("# Default Storage Parameters\n"
+        "db_host: localhost:27017\n"
+        "data_dir: ~/.convokit/saved-corpora\n"
+        "default_storage_mode: mem")
+
+class StorageManager:
+    def __init__(self,
+                 storage_type: str,
+                 db_host: Optional[str] = None,
+                 data_dir: Optional[str] = None,
+                 corpus_id: Optional[str] = None,
+                 version: Optional[str] = '0',
+                 in_place: Optional[bool] = True):
+        """
+        Object to manage data storage. 
+
+        :param storage_type: either 'mem' or 'db'
+        :param db_host: name of the DB host
+        :param data_dir: path to the directory containing Mem Corpora. If left 
+            unspecified, will use the data_dir specified in ~/.convokit/config.yml
+        :param corpus_id: id of the corpus this StorageManager will connect to.
+        :param version: the version to load from e.g. '0' or '1.0.1'
+        :param in_place: if True, directly connects to the specified version, else
+            this increments the version number so that the original copy
+            of the data is left unmodified.
+        
+        :ivar storage_type: either 'mem' or 'db'
+        :ivar index: ConvoKitIndex to track types of metadata stored in this StorageManager
+        :ivar corpus_id: id of the corpus this StorageManager is connected to.
+        :ivar CollectionMapping: class constructor to use to store collections of items.
+            Either a DBCollectionMapping or MemCollectionMapping
+        :ivar ItemMapping: class constructor to use to store data for items in the collections.
+            Either a DBDocumentMapping or MemDocumentMapping
+        """
+        if storage_type not in ['mem', 'db', None]:
+            raise ValueError(
+                f'storage_type must be "mem", "db" or None; got "{storage_type} instead"'
+            )
+        # Use defaults from config file or specified values if provided
+        self.config_fullpath = os.path.expanduser('~/.convokit/config.yml')
+        config = read_or_create_config(self.config_fullpath)
+
+        tmp_mode = os.environ.get('CONVOKIT_STORAGE_MODE')
+        if tmp_mode in ['mem', 'db']:
+            config['default_storage_mode'] = tmp_mode
+
+        if storage_type is None:
+            storage_type = config['default_storage_mode']
```

Comment:
I get the design pattern here, but IMO, configuring the `storage_type` to use should be the responsibility of whichever class is using the `StorageManager`, i.e. the `Corpus`

Reply:
hmm, could you point me to where in `corpus.py` this is handled? I'm looking at L82 and it looks like the `storage_type` is passed directly without checks.

Reply:
Lines 48-51 check if not mem, db, or None. Lines 56-58 check for the environment storage mode. Thus, if mem or db isn't specified for corpus, lines 60-61 will utilize the environment storage mode from the lines 56-58.

Reply:
@oscarso2000 Yeah, I know it's handled in `StorageManager`. My point is that this logic for fetching the appropriate storage mode should be the responsibility of `Corpus` or whichever class wants to rely on `StorageManager`, but that's just me being nitpicky (: Consider this resolved.